// napajs/test/memory-test.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
var napa = require("../lib/index");
var assert = require("assert");
describe('napajs/memory', function () {
    var napaZone = napa.zone.create('zone5');
    describe('Handle', function () {
        it('#isEmpty', function () {
            assert(napa.memory.isEmpty([0, 0]));
            assert(!napa.memory.isEmpty([1, 2]));
        });
    });
    describe('Shareable', function () {
        it('#isShareable', function () {
            assert(napa.memory.isShareable(napa.memory.crtAllocator));
            assert(!napa.memory.isShareable("hello world"));
            assert(!napa.memory.isShareable(1));
            assert(!napa.memory.isShareable({ a: 1 }));
        });
    });
    describe('Allocators', function () {
        it('@node: crtAllocator', function () {
            var handle = napa.memory.crtAllocator.allocate(10);
            assert(!napa.memory.isEmpty(handle));
            napa.memory.crtAllocator.deallocate(handle, 10);
        });
        it('@napa: crtAllocator', function () {
            napaZone.execute('./napa-zone/test', "crtAllocatorTest");
        });
        it('@node: defaultAllocator', function () {
            var handle = napa.memory.defaultAllocator.allocate(10);
            assert(!napa.memory.isEmpty(handle));
            napa.memory.defaultAllocator.deallocate(handle, 10);
        });
        it('@napa: defaultAllocator', function () {
            napaZone.execute('./napa-zone/test', "defaultAllocatorTest");
        });
        it('@node: debugAllocator', function () {
            var allocator = napa.memory.debugAllocator(napa.memory.defaultAllocator);
            var handle = allocator.allocate(10);
            assert(!napa.memory.isEmpty(handle));
            allocator.deallocate(handle, 10);
            var debugInfo = JSON.parse(allocator.getDebugInfo());
            assert.deepEqual(debugInfo, {
                allocate: 1,
                allocatedSize: 10,
                deallocate: 1,
                deallocatedSize: 10
            });
        });
        it('@napa: debugAllocator', function () {
            napaZone.execute('./napa-zone/test', "debugAllocatorTest");
        });
    });
});
