// napajs/lib/zone/zone.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
/// <summary> Default ZoneSettings </summary>
exports.DEFAULT_SETTINGS = {
    /// <summary> Set default workers to 2. </summary>
    workers: 2
};
/// <summary> Represent option to transport arguments in zone.execute. </summary>
var TransportOption;
(function (TransportOption) {
    /// <summary> transport.marshall/unmarshall will be done by `napajs` automatically.
    /// This is the most common way, but may not be performance optimal with objects
    /// that will be shared in multiple zone.execute.
    /// </summary>
    TransportOption[TransportOption["AUTO"] = 0] = "AUTO";
    /// <summary> transport.marshall/unmarshall will be done by user manually. </summary>
    TransportOption[TransportOption["MANUAL"] = 1] = "MANUAL";
})(TransportOption = exports.TransportOption || (exports.TransportOption = {}));
/// <summary> Default execution options. </summary>
exports.DEFAULT_CALL_OPTIONS = {
    /// <summary> No timeout. </summary>
    timeout: 0,
    /// <summary> Set argument transport option to automatic. </summary>
    transport: TransportOption.AUTO
};
