// napajs/lib/transport.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var transportable_1 = require("./transport/transportable");
exports.TransportableObject = transportable_1.TransportableObject;
exports.isTransportable = transportable_1.isTransportable;
exports.cid = transportable_1.cid;
__export(require("./transport/transport"));
var functionTransporter = require("./transport/function-transporter");
var binding = require('./binding');
/// <summary> Create a transport context. </summary>
function createTransportContext(owning, handle) {
    if (owning === void 0) { owning = true; }
    if (handle === void 0) { handle = undefined; }
    return new binding.TransportContextWrap(owning, handle);
}
exports.createTransportContext = createTransportContext;
exports.saveFunction = functionTransporter.save;
exports.loadFunction = functionTransporter.load;
